sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,1045 @@
+         * Signals operator termination with completion.
+         * <ul>
+         *     <li>Invoked by the upstream when it completes, or</li>
+         *     <li>Invoked by the the drain-loop when it detects first receiver signaled terminal completion and retry
+         *     is disabled (i.e., NULL_RETRY_POLICY is set).</li>
+         * </ul>
+         */
+        @Override
+        public void onComplete() {
+            if (done) {
+                return;
+            }
+
+            done = true;
+            final String logMessage;
+            if (retryPolicy == NULL_RETRY_POLICY) {
+                logMessage = "Terminal completion signal from Upstream|Receiver arrived at MessageFlux.";
+            } else {
+                logMessage = "Terminal completion signal from Upstream arrived at MessageFlux.";
+            }
+            mediatorHolder.withReceiverInfo(logger.atWarning()).log(logMessage);
+            drain(null);
+        }
+
+        /**
+         * Invoked by the downstream message subscriber to signal the demand for messages. Whatever has been
+         * requested can be sent downstream, so only signal the demand for what can be safely handled.
+         * No messages will be sent downstream until the demand is signaled.
+         *
+         * @param n the number of messages to send to downstream.
+         */
+        @Override
+        public void request(long n) {
+            if (Operators.validate(n)) {
+                Operators.addCap(REQUESTED, this, n);
+                drain(null);
+            }
+        }
+
+        /**
+         * Invoked by downstream to signal termination by cancellation.
+         */
+        @Override
+        public void cancel() {
+            if (cancelled) {
+                return;
+            }
+
+            cancelled = true;
+            mediatorHolder.withReceiverInfo(logger.atWarning())
+                .log("Downstream cancellation signal arrived at MessageFlux.");
+            // By incrementing wip, indicate the active drain-loop that there is cancel signal to handle,
+            if (WIP.getAndIncrement(this) == 0) {
+                // but it is identified that there is no active drain-loop; hence immediately react to cancel.
+                upstream.cancel();
+                mediatorHolder.freeze();
+                // wip increment also placed a tombstone on the drain-loop, so further drain(..) calls are nop.
+            }
+        }
+
+        /**
+         * Invoked by the new mediator when it is ready to be used. A mediator constructed in 'onNext' moves
+         * to ready state when its backing receiver is active.
+         *
+         * @param updateDispositionFunc the function to disposition messages from mediator's backing receiver.
+         */
+        void onMediatorReady(BiFunction<String, DeliveryState, Mono<Void>> updateDispositionFunc) {
+            retryAttempts.set(0);
+            parent.onNextUpdateDispositionFunction(updateDispositionFunc);
```

Comment:
Do we need this method that takes an updateDispositionFunc? I only see it in one place and I feel like it makes the chain harder to follow. I tried tof ind other callees of this bifunction but could not.

Reply:
Since MessageFlux abstracts rolling over multiple Receivers, the MessageFlux needs to be notified with the function on the current receiver allowing to disposition, so the notification flow: "Mediator (Receiver) -> RecoverableReactorReceiver -> MessageFlux". I’ve some ideas to make it more readable, will update this.